### ISSUE TITLE
Add CodeIntel as a completer for Go, PHP, Perl, Ruby, and Javascript

### ISSUE THREAD

**Issue by cweagans** [Valloric/ycmd#96]
CodeIntel is the code completion engine from Komodo Edit, and it supports a ton of languages and does a pretty good job at most of them. I'd love to see it integrated with ycmd, and I may end up paying somebody to write the completers for it or I may spend the time to learn how to make it happen myself.

In [my fork](https://github.com/cweagans/ycmd), I've added CodeIntel as a third party library. I'm not sure how @Valloric will want to go about using it as a completer for so many different languages, so I'm opening this early with the hopes of getting architectural feedback on it.

**Comment by vheon** [comment]
I'm not thrilled about the idea of using one big tool for autocompleting every language some of which are already integrated into ycmd, like python. But that is just me.

**Comment by cweagans** [comment]
Clang is already kind of used for that? Also, we don't have to use CodeIntel for Python (and we shouldn't, IMO. Jedi is better). Using it to add as many languages as possible seems to make sense, though.

Go could also be omitted and provided by Gocode instead.

**Comment by vheon** [comment]
Yeah, but you see where I'm going? You are saying: "we should use CodeIntel so we could bring more languages... but not for python because Jedi is better... and not for go because geocode is more appropriate..." and we could go on, like "and not for javascript because tern.js does already a great job at that..." and there you go that from the list of the good completion of CodeIntel we remain with 2-3 languages. Now I won't say that having the completion system for those 2-3 languages would be bad, but if we consider the effort that the developer of CodeIntel will put into the project will be shared by all those language and therefore would be less than if we had a completion system for each language. Again this is just my opinion, but considering that @Valloric already said that he would not prefer having the user configure which autocompletion system use for a specific language but deciding upfront which will be the better option for the user (so a sane default) I would go with caution before committing the project to a completer, thats all.

**Comment by cweagans** [comment]
Yeah, I see what you're saying. However, if we're going to add CodeIntel for 2-3 languages that aren't provided by anything else, we might as well also add it for languages that *are* provided by other things that we don't have yet. If Gocode is a better option in the future, then we simply switch. There's also no reason we can't have both CodeIntel and gocode for Go completion. We can query both completers and merge the results.

At the end of the day, I have no strong opinions about what languages from CodeIntel we actually use, but I really want to see PHP added. There are a ton of people out there using languages that aren't provided by ycmd, and if ycmd provides autocomplete for them, people are more likely integrate ycmd with more text editors and everyone benefits.

**Comment by Valloric** [comment]
> There's also no reason we can't have both CodeIntel and gocode for Go completion.

There's a very good reason, and that's presenting to the user a system that Just Works. No "hey pick a semantic engine you like" stuff. I don't want to have several possible semantic engines for any single language; that way lies madness. Hard to maintain and an explosion of complexity.

I'm not saying I'm against CodeIntel for semantic engines for languages we don't have semantic support now, but I am saying that if at some point we get a better semantic engine for a language supported by CodeIntel, then CodeIntel would be turned off for that language.

I also think that while Tern.js is probably better at JS completion than CodeIntel, if CodeIntel is "good enough" than we can plug that in until we have Tern.js support (which might take a while). But only if it really is good enough; I don't want to connect semantic engines that aren't up to certain level of quality. Just recently I looked into plugging in Racer for Rust completion, but it's too young and needs more work before it's acceptable as a ycmd backend.

In other words, I'd rather only have identifier completion for a language than shitty semantic completion.

**Comment by Valloric** [comment]
@cweagans The next step would be to send a [WIP] pull request to ycmd so that I can take a look at what would integrating CodeIntel entail. I can't promise this will be merged; I don't know how big of a maintenance burden supporting CodeIntel will be until I see a PR pulling it in.

**Comment by cweagans** [comment]
@Valloric: Okay, that's fair. I just remember reading something about multiple semantic engines getting queried simultaneously sometimes. In any case, I see where you're coming from. I know for sure that CodeIntel is very good for PHP, so maybe we can start there.

Do you want all of the codeintel integration in a PHP completer directory (`ycmd/ycmd/completers/php`) or somewhere else (considering that codeintel can do autocompletion for a bunch of other languages that we might end up using at some point)?

**Comment by oblitum** [comment]
@Valloric While were're at it (sorry for hijacking the discussion), I've been thinking whether it would be better if ycmd was engine-pluggable like our own Vim plugins. I feel embedding too much (N-engines) in the backend is turning YCM/ycmd too heavy weight and things should become more heavy as time passes by. For example, YCM is the only plugin I'm unable to clone by default using Vim-Plug, with my connection at last, there's a ton of submodules to clone, but I don't really care about C# support for example.

**Comment by oblitum** [comment]
@Valloric by the way, that's just food for thought.

**Comment by Valloric** [comment]
@oblitum In theory I might agree with you, but I think that ship has sailed.

@cweagans Start by setting it up for PHP and we can refactor it to support other languages later if experience with PHP shows promise.

**Comment by cweagans** [comment]
:+1:

**Comment by vheon** [comment]
>  For example, YCM is the only plugin I'm unable to install using Vim-Plug by default, I have to clone by hand or change some timeout, with my connection at last. There's a ton of submodules to clone, but I don't really care about C# support for example.

@oblitum I'm in the same situation :(
@Valloric Could we have something were we explicitly tell that we don't want something? So by default we leaves things as they are but if I say something like `install.sh --no-omnisharp-completer` we don't download the submodule for the C# completer?

**Comment by cweagans** [comment]
Sounds like a different issue to me...

**Comment by oblitum** [comment]
@vheon my ideal would ycmd as completion framework/standard for engines that claim to extend it but that live in independent repos, so that I would need to list then also in my Vim-Plug plugin list.

**Comment by oblitum** [comment]
@vheon well, more or less like I said (it's not just a Vim plugin anymore, I know).

**Comment by cweagans** [comment]
@Valloric CodeIntel is now available via `pip install codeintel`. I'm working on getting docs about the protocol, how to connect, etc. In the mean time, is there a particular way that you'd recommend for pulling in the codeintel daemon?

**Comment by vheon** [comment]
@cweagans where is the main repo? the current way of importing semantic engines is to add a submodule in the `third_party` directory as you can see [here](https://github.com/Valloric/ycmd/tree/master/third_party)

**Comment by cweagans** [comment]
I see that. I haven't been able to find a repo for the codeintel that's installed through pip. This is the only thing that I've seen: https://pypi.python.org/pypi/CodeIntel

**Comment by micbou** [comment]
We won't use CodeIntel as we prefer semantic engines dedicated to a language (or a family of languages) since they are likely to have more features for the language they are targeting than CodeIntel. But even if we were willing to use this tool, we couldn't because [its license is not compatible with the GPL](https://www.gnu.org/licenses/license-list.html#MPL). So that's it for CodeIntel.
